### PR TITLE
FD refactor + sub device support

### PR DESCRIPTION
**Pull request by tt-aho**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/13655

### Problem description
Various aspects of FD/runtime need to be updated to support multiple sub devices.
SubDevice: Container of the various cores we want to group together from a device
SubDeviceManager: Container of the state/objects that represent a configuration of the device that is split into SubDevices.

### What's changed
Support multiple dispatch entries for worker->dispatch sync

Update dispatch d/s to have a semaphore per dispatch entry to enable syncing on specific worker counts

Update LaunchMessageRingBufferState and WorkerConfigBufferMgr to be tracked per sub_device

Update various FD commands to support syncing on multiple sub devices:
- ERB, EWB, ERE, Synchronize/finish takes in a list of sub devices for blocking/issuing waits on. Will wait on all sub-devices if none are provided
- Trace will track only specific sub devices used
- EP currently only supports one sub-device

Update allocator to support taking in a CoreRangeSet for banks, instead of assuming a rectangular grid

Add support for splitting a device into multiple SubDevices, as well and maintaining different SubDeviceManager configurations, owned by device

Add basic tests to validate sub-device support

I will be continuing to add/updates tests in the near future, but would like to get this merged since this touches a lot of files/state to avoid dealing with conflicts.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11823639393
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by mo-tenstorrent — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 1]
```diff
@@ -1581,11 +1600,22 @@ void EnqueueRecordEventCommand::process() {
     HugepageDeviceCommand command_sequence(cmd_region, cmd_sequence_sizeB);
 
     CoreType dispatch_core_type = dispatch_core_manager::instance().get_dispatch_core_type(this->device->id());
-    uint32_t dispatch_message_addr = dispatch_constants::get(
+    uint32_t dispatch_message_base_addr = dispatch_constants::get(
         dispatch_core_type).get_device_command_queue_addr(CommandQueueDeviceAddrType::DISPATCH_MESSAGE);
 
+    uint32_t last_index = this->expected_num_workers_completed.size() - 1;
```
Comment (pgkeller) [inline tt_metal/impl/dispatch/command_queue.cpp:1606]: don't follow what this is doing.
code here will conflict w/ the kernel bin ring buffer async changes

**Review by pgkeller — COMMENTED** [review]
only got part way through, will review again later

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 1 above)
Comment (tt-aho) [reply tt_metal/impl/dispatch/command_queue.cpp:1609]: The code change here is that there are now multiple worker completion counters. Previously we only had one counter for the whole chip, and we would only need a single wait cmd. We now enqueue a wait cmd per counter (todo to take in a list of which counters to wait on), and we specialize the last iteration because we only need a write barrier on the dispatch core once at the end, and is not needed per wait cmd.

What would be the conflict with the ring buffer changes?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 2]
```diff
@@ -1581,11 +1600,22 @@ void EnqueueRecordEventCommand::process() {
...
+    // We only need the write barrier for the last wait cmd
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/impl/dispatch/command_queue.cpp:1610]: Possible. I didn't commonize it since they use different fns `add_dispatch_wait` vs `add_dispatch_wait_with_prefetch_stall` and could potentially take different args, and wanted to avoid adding a wrapper fn where we'd need to match default args with the actual fns to keep same functionality.

File: tt_metal/impl/device/device.cpp [hunk 3]
```diff
@@ -3102,6 +3126,29 @@ void Device::check_allocator_is_initialized() const {
     }
 }
 
+void Device::reset_num_sub_devices(uint32_t num_sub_devices) {
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/impl/device/device.cpp:3121]: Yes. I tested locally by inserting it into the setup/teardown process for device and running a few tests to verify functionality, but nothing uses it in the actual pr.

File: tt_metal/impl/dispatch/command_queue_interface.hpp [hunk 4]
```diff
@@ -64,8 +65,12 @@ struct dispatch_constants {
         return *inst;
     }
 
+    using prefetch_q_entry_type = uint16_t;
+
     static constexpr uint8_t MAX_NUM_HW_CQS = 2;
-    typedef uint16_t prefetch_q_entry_type;
+    static constexpr uint32_t DISPATCH_MESSAGE_ENTRIES = 16;
```
Comment (aliuTT) [inline tt_metal/impl/dispatch/command_queue_interface.hpp:71]: Is this the max number of sub devices? It seems like it's used across the stack to represent that. It may be clearer to have another `MAX_NUM_SUBDEVICES = DISPATCH_MESSAGE_ENTRIES`.

**Review by aliuTT — APPROVED** [review]

File: tt_metal/impl/dispatch/command_queue_interface.hpp (hunk 4 above)
Comment (cfjchu) [reply tt_metal/impl/dispatch/command_queue_interface.hpp:71]: +1 I was looking at device changes and was wondering what `DISPATCH_MESSAGE_ENTRIES` was supposed to be.

Comment (tt-aho) [reply tt_metal/impl/dispatch/command_queue_interface.hpp:71]: I've added a new static constexpr in Device, `static constexpr uint32_t MAX_NUM_SUB_DEVICES = dispatch_constants::DISPATCH_MESSAGE_ENTRIES;`, and updated the relevant use cases in device files to use this instead of `DISPATCH_MESSAGE_ENTRIES`. `DISPATCH_MESSAGE_ENTRIES` is still used where appropriate. Let me know if you think it should be switched elsewhere.

File: tt_metal/impl/device/device.cpp [hunk 5]
```diff
@@ -3405,6 +3553,113 @@ void Device::generate_device_headers(const std::string &path) const
...
+}
+
+std::optional<SubDeviceManagerId> Device::get_active_sub_device_manager_id() const {
+    return this->active_sub_device_manager_id_;
+}
+
+SubDeviceManagerId Device::create_sub_device_manager(tt::stl::Span<const SubDevice> sub_devices, DeviceAddr local_l1_size) {
+    static std::atomic<SubDeviceManagerId> sub_device_manager_id = 0;
+    TT_FATAL(!this->using_slow_dispatch(), "Using sub device managers is unsupported with slow dispatch");
+    auto [sub_device_manager, _] = this->sub_device_managers_.insert_or_assign(sub_device_manager_id, std::make_unique<detail::SubDeviceManager>(sub_devices, local_l1_size, this));
```
Comment (patrickroberts) [inline tt_metal/impl/device/device.cpp:3626]: Seeing as this appears to be the only time the `SubDeviceManager` is ever constructed, is it necessary to keep the constructor overload that move-constructs a vector? I don't see it being used anywhere.

File: tt_metal/tt_metal.cpp [hunk 6]
```diff
@@ -1137,37 +1150,38 @@ std::unique_ptr<GlobalSemaphore> CreateGlobalSemaphore(
...
-std::shared_ptr<Buffer> CreateBuffer(const ShardedBufferConfig &config) {
-    return Buffer::create(
-        config.device,
-        config.size,
-        config.page_size,
-        config.buffer_type,
-        config.buffer_layout,
-        config.shard_parameters,
-        std::nullopt);
+std::shared_ptr<Buffer> CreateBuffer(const InterleavedBufferConfig &config, std::optional<DeviceAddr> address, std::optional<uint32_t> sub_device_id) {
```
Comment (patrickroberts) [inline tt_metal/tt_metal.cpp:1153]: Please don't combine these overloads, I'd rather avoid the unnecessary `std::optional` branching here, where the logical was previously unconditional.

Comment (tt-aho) [reply tt_metal/tt_metal.cpp:1153]: Do you have a suggestion for avoiding this? I ended up combining them because the uint64_t address and The uint32_t sub_device_id are both integer type, so it was ambiguous to overload.

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 7]
```diff
@@ -2853,7 +3039,8 @@ void EnqueueReadBuffer(
     CommandQueue& cq,
     std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>> buffer,
     void* dst,
-    bool blocking) {
+    bool blocking,
+    tt::stl::Span<const uint32_t> sub_device_ids) {
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/impl/dispatch/command_queue.cpp:3113]: Yes, will fix. I messed up when I was splitting my commits between the FD refactor and the sub-device implementation

File: tt_metal/impl/device/device.cpp [hunk 8]
```diff
@@ -3096,25 +3143,83 @@ uint32_t Device::get_noc_multicast_encoding(uint8_t noc_index, const CoreRange&
...
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
+        return allocator;
+    } else {
+        if (!this->allocator_) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
+        return this->allocator_;
+    }
+}
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/impl/device/device.cpp:3205]: I'll replace it

File: tt_metal/impl/device/device.cpp [hunk 9]
```diff
@@ -3096,25 +3143,83 @@ uint32_t Device::get_noc_multicast_encoding(uint8_t noc_index, const CoreRange&
...
+        auto& allocator = sub_device_manager->sub_device_allocator(*sub_device_id);
+        if (!allocator) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
+    } else {
+        if (!this->allocator_) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
     }
 }
```
[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/impl/device/device.cpp:3187]: Yes, will separate

File: tt_metal/tt_metal.cpp (hunk 6 above)
Comment (patrickroberts) [reply tt_metal/tt_metal.cpp:1153]: Make `sub_device_id` a strong type. `DeviceAddr` should have been a strong type too. Making them aliases for numeric types is what leads to completely avoidable issues like this.

File: tt_metal/impl/device/device.cpp [hunk 10]
```diff
@@ -2895,6 +2934,12 @@ bool Device::close() {
         if (hw_command_queue->manager.get_bypass_mode()) {
             hw_command_queue->record_end();
         }
+    }
+    this->clear_loaded_sub_device_manager();
+    for (auto sub_device_manager = this->sub_device_managers_.begin(); sub_device_manager != this->sub_device_managers_.end();) {
+        this->remove_sub_device_manager((sub_device_manager++)->first);
+    }
```
[the remark this answers is not in the document]
Comment (patrickroberts) [reply tt_metal/impl/device/device.cpp:2968]: https://en.cppreference.com/w/cpp/container/unordered_map/erase:
> References and iterators to the erased elements are invalidated. **Other iterators and references are not invalidated.** (emphasis added)

[the remark this answers is not in the document]
Comment (tt-aho) [reply tt_metal/impl/device/device.cpp:2968]: This does not appear to work.
Ex. In one of my tests I see the original loop remove ids 2, 1, 0.
Replacing this code with the suggestion I see it clear 2, then a random value causing it to error out.

**Review by patrickroberts — APPROVED** [review]

File: tt_metal/tt_metal.cpp [hunk 11]
```diff
@@ -1139,14 +1154,37 @@ std::unique_ptr<GlobalSemaphore> CreateGlobalSemaphore(
...
     return Buffer::create(
-        config.device, config.size, config.page_size, config.buffer_type, config.buffer_layout, std::nullopt, std::nullopt);
+        config.device,
+        config.size,
+        config.page_size,
+        config.buffer_type,
+        config.buffer_layout,
+        std::nullopt,
+        std::nullopt,
+        std::nullopt);
```
Comment (abhullar-tt) [inline tt_metal/tt_metal.cpp:1164]: this can probably be cleaned up in a separate PR but could we add these extra params to the config? I think some are common between Interleaved and Sharded configs but having a bunch of nullopts and then sub device ids looks awkward. Could you create an issue for this

**Review by abhullar-tt — APPROVED** [review]

File: tt_metal/tt_metal.cpp (hunk 11 above)
Comment (tt-aho) [reply tt_metal/tt_metal.cpp:1164]: Yes, I'll file an issue for this. The three nullopts are shard_parameters, bottom_up (this can probably go into the config), and sub_device_id. Do we want sub_device_id in the config as well, or does it make more sense as it is right now where host_api takes it as a separate argument like device addr?

Comment (abhullar-tt) [reply tt_metal/tt_metal.cpp:1164]: since it isn't really parameter of how to set up the buffer I think its fine for it to go outside the config
